Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -0,0 +1,828 @@
+                System.arraycopy(currentClusterLabels, 0, hierarchyLevelLabels, 0, currentClusterLabels.length);
+                newCluster.setHierarchyLevel(lineCount);
+                hierarchy.put(lineCount, hierarchyLevelLabels);
+            }
+
+            if (newClusters.isEmpty())
+                nextLevelSignificant = false;
+            else
+                nextLevelSignificant = true;
+        }
+        return clusters;
+    }
+
+    /**
+     * Propagates stability and lowest child death level from each child cluster to its parent cluster in the tree.
+     * This step is required before calling prominent clusters can be identified or outlier scores can be calculated.
+     * @param clusters A list of clusters representing the cluster tree.
+     */
+    private static void propagateTree(List<HdbscanCluster> clusters) {
+        TreeMap<Integer, HdbscanCluster> clustersToExamine = new TreeMap<>();
+        BitSet addedToExaminationList = new BitSet(clusters.size());
+
+        // Find all leaf clusters in the cluster tree:
+        for (HdbscanCluster cluster : clusters) {
+            if (cluster != null && !cluster.hasChildren()) {
+                clustersToExamine.put(cluster.getLabel(), cluster);
+                addedToExaminationList.set(cluster.getLabel());
+            }
+        }
+
+        // Iterate through every cluster, propagating stability from children to parents:
+        while (!clustersToExamine.isEmpty()) {
+            HdbscanCluster currentCluster = clustersToExamine.pollLastEntry().getValue();
+            currentCluster.propagate();
+
+            if (currentCluster.getParent() != null) {
+                HdbscanCluster parent = currentCluster.getParent();
+
+                if (!addedToExaminationList.get(parent.getLabel())) {
+                    clustersToExamine.put(parent.getLabel(), parent);
+                    addedToExaminationList.set(parent.getLabel());
+                }
+            }
+        }
+    }
+
+    /**
+     * Produces the cluster labels using cluster stability values.
+     *
+     * @param hierarchy The HDBSCAN* hierarchy.
+     * @param clusters A list of clusters representing the cluster tree.
+     * @param numPoints The number of points in the dataset.
+     * @return A list containing every point's cluster label.
+     */
+    private static List<Integer> findProminentClusters(Map<Integer, int[]> hierarchy, List<HdbscanCluster> clusters,
+                                                      int numPoints) {
+
+        // Take the list of propagated clusters from the root cluster:
+        List<HdbscanCluster> solution = clusters.get(1).getPropagatedDescendants();
+
+        List<Integer> clusterLabels = new ArrayList<>(Collections.nCopies(numPoints, OUTLIER_NOISE_CLUSTER_LABEL));
+
+        // Store all the levels at which to find the birth points for the flat clustering:
+        TreeMap<Integer, List<Integer>> significantLevels = new TreeMap<>();
```

Comment:
This also seems like it could be a priority queue, though it might need to be a queue of a `Pair<Integer,List<Integer>>`.

Reply:
I wasn't sure that for this case a PriorityQueue<Pair<Integer, List<Integer>>> will bring a significant improvement over the TreeMap. Any link or literature to share?

Reply:
It looks like `TreeMap` implementation traverses the tree to get the last entry (https://github.com/openjdk/jdk11u/blob/master/src/java.base/share/classes/java/util/TreeMap.java#L2134), so the poll on the priority queue might be better. But it doesn't matter too much.

Reply:
Thanks for providing this reference (that I should have checked myself) - performance always matters and O(1) is certainly more desirable than O(log n). I tried changing this with a `PriorityQueue<Pair<Integer,List<Integer>>> pq = new PriorityQueue<>(Comparator.comparing(Pair::getA));` but was finding the `contains` method from the `PriorityQueue` a little awkward, to check if the queue contains a `Pair<Integer,List<Integer>>` so that `List<Integer>` can be appended to.  I think we should stick with TreeMap here for now.

Reply:
Ok, that's fine then.